### PR TITLE
Custom bin

### PR DESCRIPTION
**Pull request by elfet**
| Q             | A
| ------------- | ---
| Bug fix?      | No
| New feature?  | Yes
| BC breaks?    | No
| Deprecations? | Yes `composer_command`
| Fixed tickets | #548 #607 #586 #606

This is my version of custom bins. Here an example:

```php
task('default', function () {
    run('{{bin/php}} -v');
    run('{{bin/git}} --version');
    run('{{bin/composer}} --version');
});
```
What do you think, guys?

### CODE REVIEW

**Comment by nemoinho** [comment]
All in all it looks good to me but would you really download composer if it's missing? I think that should only rise an exception and stop deployer. And maybe deployer should test for a few common locations of composer before doing that. I really don't like the idea, that a application loads extra software without warning or at least an info.

**Comment by elfet** [comment]
For now composer automatically downloaded too. If we remove this, may broke BC.

**Comment by elfet** [comment]
Also it's downloaded via `https`, so risks a minimal. I think this behavior is good for developer experience, and we should keep it. You can always install composer manually to prevent downloading it.

**Comment by nemoinho** [comment]
Yep, unfortunately, but maybe we can keep this topic  on the agenda for a next major release.

**Comment by elfet** [comment]
:+1:

File: recipe/common.php [hunk 1]
```diff
@@ -39,6 +38,30 @@
...
+});
+
+env('bin/composer', function () {
+    if (commandExist('composer')) {
+        $composer = run('which composer')->toString();
+    }
+
+    if(empty($composer)) {
+        run("cd {{release_path}} && curl -sS https://getcomposer.org/installer | php");
+        $composer = 'php composer.phar';
```
Comment (oanhnn) [inline recipe/common.php]: Please replace with `{{bin/php}} composer.phar`

Comment (elfet) [inline recipe/common.php]: :+1:

**Comment by oanhnn** [comment]
Users can void download `composer.phar` by a task upload `composer.phar` to shared directory and:
```
env('bin/composer', '{{bin/php}} {{deploy_path}}/shared/composer.phar');
```
:smile:

**Comment by elfet** [comment]
@deployphp/maintainer merge? :)

**Comment by oanhnn** [comment]
I think we will merge to both 3.1.x and 4.x

**Comment by oanhnn** [comment]
@elfet I merged branch `custom_bin` and `master` to [`oanhnn/4.x`](https://github.com/oanhnn/deployer/tree/4.x). Now you can merge :smile:

**Comment by elfet** [comment]
I think merge master to 4.x too :)

**Comment by oanhnn** [comment]
Yeah, I merged :smile:

**Comment by rvanlaak** [comment]
@antonmedv would it be easy to change the php bin per host?

Our use-case: we would like to start experimenting with PHP7.4 without affecting the other applications on that host.
The vhost was changed to let Apache serve PHP7.4, but deployment fails because CLI is on PHP7.3.

**Pull request merged**